examples/c/task_iter.c

```diff
@@ -0,0 +1,130 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+/* Copyright (c) 2020 Facebook */
+#include <argp.h>
+#include <signal.h>
+#include <stdio.h>
+#include <sys/resource.h>
+#include <bpf/bpf.h>
+#include <bpf/libbpf.h>
+#include <unistd.h>
+#include "task_iter.h"
+#include "task_iter.skel.h"
+
+static struct env {
+	bool verbose;
+} env;
+
+static int libbpf_print_fn(enum libbpf_print_level level, const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static volatile bool exiting = false;
+
+static void sig_handler(int sig)
+{
+	exiting = true;
+}
+
+void print_task_state(__u32 state)
+{
+	// taken from
+	// https://elixir.bootlin.com/linux/latest/source/include/linux/sched.h#L85
+	// there are a lot more states not covered here but these are common ones
+	switch (state) {
+	case 0x0000:
+		printf("State: RUNNING\n");
+		break;
+	case 0x0001:
+		printf("State: INTERRUPTIBLE\n");
+		break;
+	case 0x0002:
+		printf("State: UNINTERRUPTIBLE\n");
+		break;
+	case 0x0200:
+		printf("State: WAKING\n");
+		break;
+	case 0x0400:
+		printf("State: NOLOAD\n");
+		break;
+	case 0x0402:
+		printf("State: IDLE\n");
+		break;
+	case 0x0800:
+		printf("State: NEW\n");
+		break;
+	default:
+		printf("Unknown Task state\n");
+	}
+}
+
+int main(int argc, char **argv)
+{
+	struct task_iter_bpf *skel;
+	int err;
+	int get_tasks_fd;
+
+	/* Set up libbpf errors and debug info callback */
+	libbpf_set_print(libbpf_print_fn);
+
+	/* Cleaner handling of Ctrl-C */
+	signal(SIGINT, sig_handler);
+	signal(SIGTERM, sig_handler);
+
+	/* Load and verify BPF application */
+	skel = task_iter_bpf__open();
+	if (!skel) {
+		fprintf(stderr, "Failed to open and load BPF skeleton\n");
+		return 1;
+	}
+
+	/* Load & verify BPF programs */
+	err = task_iter_bpf__load(skel);
```

Comment:
combine open and load with `task_iter_bpf__open_and_load()`?

Reply:
oooo TIL